test/integration/http_server_action_test.lua

```diff
@@ -0,0 +1,79 @@
+local fio = require('fio')
+local fiber = require('fiber')
+local socket = require('socket')
+local t = require('luatest')
+local http_client = require('http.client')
+local http_server = require('http.server')
+
+local helpers = require('test.helpers')
+
+local g = t.group()
+
+g.before_test('test_stop_basic', function()
+    g.httpd = helpers.cfgserv()
+    g.httpd:start()
+end)
+
+g.before_test('test_stop_with_open_connection', function()
+    g.httpd = helpers.cfgserv()
+    g.httpd:start()
+end)
+
+g.after_test('test_stop_basic', function()
+    helpers.retrying({}, function()
+        local conn = socket.tcp_connect(helpers.base_host, helpers.base_port)
+        return conn ~= nil
+    end)
+    fio.rmdir(helpers.path)
+end)
+
+g.after_test('test_stop_with_open_connection', function()
+    helpers.retrying({}, function()
+        local r = http_client.request('GET', helpers.base_uri)
+        return r == nil
+    end)
+    fio.rmdir(helpers.path)
+end)
+
+g.test_stop_basic = function()
+    local httpd = g.httpd
+    httpd:route({
+        method = 'GET',
+        path = '/test/json',
+    }, 'test#json')
+
+    local resp = httpd:stop()
+    t.assert_not_equals(resp, nil)
+    t.assert_type(resp, 'table')
```

Comment:
What's in this response table?

Reply:
There is a huge table with information about http server. I'm not sure that we need to check fields in this table. It is enough checking it is not `nil`. What do you think?

```lua
{
    cache = {ctx = {}, static = {}, tpl = {}},
    helper = function: 0x40ec64b0,
    helpers = {helper_title = function: 0x41253358, url_for = function: 0x41fa60d8},
    hook = function: 0x40ec64d0,
    hooks = {},
    host = "127.0.0.1",
    iroutes = {star = 10, test = 1},
    is_run = false,
    match = function: 0x40ec6490,
    options = {
        app_dir = "./test",
        cache_controllers = true,
        cache_static = true,
        cache_templates = true,
        charset = "utf-8",
        display_errors = false,
        handler = function: 0x415fe038,
        header_timeout = 100,
        log_errors = false,
        log_requests = false,
        max_header_size = 4096,
    },
    port = 12345,
    route = function: 0x40ec6670,
    routes = {
        {
            match = "^/abc/([^/]-)/([^/]-)/$",
            method = "ANY",
            name = "test",
            path = "/abc/:cde/:def",
            stash = {"cde", "def"},
            sub = function: 0x41251470,
            url_for = function: 0x40ec64f0,
        },
        {
            match = "^/abc/$",
            method = "ANY",
            path = "/abc",
            stash = {},
            sub = function: 0x41251810,
            url_for = function: 0x40ec64f0,
        },
        {
            action = "action",
            controller = "module.controller",
            match = "^/ctxaction/$",
            method = "ANY",
            path = "/ctxaction",
            stash = {},
            sub = function: 0x40ec6510,
            url_for = function: 0x40ec64f0,
        },
        {
            action = "absent",
            controller = "module.controller",
            match = "^/absentaction/$",
            method = "ANY",
            path = "/absentaction",
            stash = {},
            sub = function: 0x40ec6510,
            url_for = function: 0x40ec64f0,
        },
        {
            action = "action",
            controller = "module.absent",
            match = "^/absent/$",
            method = "ANY",
            path = "/absent",
            stash = {},
            sub = function: 0x40ec6510,
            url_for = function: 0x40ec64f0,
        },
        {
            match = "^/abc/([^/]-)/$",
            method = "ANY",
            path = "/abc/:cde",
            stash = {"cde"},
            sub = function: 0x41252280,
            url_for = function: 0x40ec64f0,
        },
        {
            match = "^/abc_([^/]-)/$",
            method = "ANY",
            path = "/abc_:cde_def",
            stash = {"cde_def"},
            sub = function: 0x41252590,
            url_for = function: 0x40ec64f0,
        },
        {
            match = "^/abc[-]([^/]-)[-]def/$",
            method = "ANY",
            path = "/abc-:cde-def",
            stash = {"cde"},
            sub = function: 0x412528a0,
            url_for = function: 0x40ec64f0,
        },
        {
            match = "^/aba(.-)/$",
            method = "ANY",
            path = "/aba*def",
            stash = {"def"},
            sub = function: 0x41252be0,
            url_for = function: 0x40ec64f0,
        },
        {
            match = "^/abb(.-)/cde/$",
            method = "ANY",
            name = "star",
            path = "/abb*def/cde",
            stash = {"def"},
            sub = function: 0x41252220,
            url_for = function: 0x40ec64f0,
        },
        {
            match = "^/banners/([^/]-)/$",
            method = "ANY",
            path = "/banners/:token",
            stash = {"token"},
            sub = function: 0x415fdae0,
            url_for = function: 0x40ec64f0,
        },
        {
            file = "helper.html.el",
            match = "^/helper/$",
            method = "ANY",
            path = "/helper",
            stash = {},
            sub = function: 0x415fdae0,
            url_for = function: 0x40ec64f0,
        },
        {
            file = "test.html.el",
            match = "^/test/$",
            method = "ANY",
            path = "/test",
            stash = {},
            sub = function: 0x412533a0,
            url_for = function: 0x40ec64f0,
        },
        {
            action = "json",
            controller = "test",
            match = "^/test/json/$",
            method = "GET",
            path = "/test/json",
            stash = {},
            sub = function: 0x40ec6510,
            url_for = function: 0x40ec64f0,
        },
    },
    start = function: 0x40ec6748,
    stop = function: 0x40ec6470,
    tcp_server_f = function: 0x40bf31f0,
    url_for = function: 0x40ec6728,
}
```

I have added a check that response is for http server that we started before (port and hostname):

```diff
--- a/test/integration/http_server_action_test.lua
+++ b/test/integration/http_server_action_test.lua
@@ -37,14 +37,12 @@ end)
 
 g.test_stop_basic = function()
     local httpd = g.httpd
     local resp = httpd:stop()
     t.assert_not_equals(resp, nil)
     t.assert_type(resp, 'table')
+    -- Make sure response is for http server that we started before.
+    t.assert_equals(resp.port, helpers.base_port)
+    t.assert_equals(resp.host, helpers.base_host)
 end
 
 g.test_stop_with_open_connection = function()
```